nowriter/tmdecoding/tm_decoder_pure.go

```diff
@@ -0,0 +1,241 @@
+	}
+	return
+}
+
+func (e TMDecoderPure) DecodeFloat32(inp []byte) (f float32, bytesRead int, err error) {
+	const primitiveSize int = 4
+	if len(inp) < primitiveSize {
+		return
+	}
+	i := uint32(binary.BigEndian.Uint32(inp[:primitiveSize]))
+	f = math.Float32frombits(i)
+	bytesRead = primitiveSize
+	return
+}
+
+func (e TMDecoderPure) DecodeFloat64(inp []byte) (f float64, bytesRead int, err error) {
+	const primitiveSize int = 8
+	if len(inp) < primitiveSize {
+		return
+	}
+	i := uint64(binary.BigEndian.Uint64(inp[:primitiveSize]))
+	f = math.Float64frombits(i)
+	bytesRead = primitiveSize
+	return
+}
+
+func (e TMDecoderPure) DecodeInt8(inp []byte) (i int8, bytesRead int, err error) {
+	const primitiveSize int = 1
+	if len(inp) < primitiveSize {
+		return
+	}
+	i = int8(inp[0])
+	bytesRead = primitiveSize
+	return
+}
+
+func (e TMDecoderPure) DecodeInt16(inp []byte) (i int16, bytesRead int, err error) {
+	const primitiveSize int = 2
+	if len(inp) < primitiveSize {
+		return
+	}
+	i = int16(binary.BigEndian.Uint16(inp[:primitiveSize]))
+	bytesRead = primitiveSize
+	return
+}
+
+func (e TMDecoderPure) DecodeInt32(inp []byte) (i int32, bytesRead int, err error) {
+	const primitiveSize int = 4
+	if len(inp) < primitiveSize {
+		return
+	}
+	i = int32(binary.BigEndian.Uint32(inp[:primitiveSize]))
+	bytesRead = primitiveSize
+	return
+}
+
+func (e TMDecoderPure) DecodeInt64(inp []byte) (i int64, bytesRead int, err error) {
+	const primitiveSize int = 8
+	if len(inp) < primitiveSize {
+		return
+	}
+	i = int64(binary.BigEndian.Uint64(inp[:primitiveSize]))
+	bytesRead = primitiveSize
+	return
+}
+func (e TMDecoderPure) DecodeOctet(inp []byte) (b byte, bytesRead int, err error) {
+	const primitiveSize int = 1
+	if len(inp) < primitiveSize {
+		return
+	}
+	b = inp[0]
+	bytesRead = primitiveSize
+	return
+}
+
+func (e TMDecoderPure) DecodeOctets(inp []byte) (b []byte, bytesRead int, err error) {
+	bytesRead = len(inp)
+	if bytesRead > 0 {
+		b = make([]byte, bytesRead)
+		copy(b, inp)
+	}
+	return
+}
+
+func (e TMDecoderPure) DecodeTime(inp []byte) (t time.Time, bytesRead int, err error) {
+	const primitiveSize int = 8
+	if len(inp) < primitiveSize {
+		return
+	}
+	i := int64(binary.BigEndian.Uint64(inp[:primitiveSize]))
+	t = time.Unix(0, i)
+	bytesRead = primitiveSize
+	return
+}
+
+func (e TMDecoderPure) DecodeUint8(inp []byte) (i uint8, bytesRead int, err error) {
+	const primitiveSize int = 1
+	if len(inp) < primitiveSize {
+		return
+	}
+	i = uint8(inp[0])
+	bytesRead = primitiveSize
+	return
+}
+func (e TMDecoderPure) DecodeUint16(inp []byte) (i uint16, bytesRead int, err error) {
+	const primitiveSize int = 2
+	if len(inp) < primitiveSize {
+		return
+	}
+	i = uint16(binary.BigEndian.Uint16(inp[:primitiveSize]))
+	bytesRead = primitiveSize
+	return
+}
+
+func (e TMDecoderPure) DecodeUint16s(inp []byte) (iz []uint16, bytesRead int, err error) {
+	if len(inp) < 4 {
+		return
+	}
+	count := int(uint32(binary.BigEndian.Uint32(inp[:4])))
```

Comment:
possible overflow? like if int is `int32`

Reply:
@melekes yeah it looked weird to me too but i reviewed the tests there were already lots around 32 bits and i added some myself around border cases of 64 bit ints so i think the code looks weird but is correct by go rules. i guess it is wraparound overflow that means. anyhow i think it's no worse than before.

Reply:
i'm not sure we are solid on 32-bit nor 64-bit arch but i am happy to run 32-bit tests if that's really a requirement.